### DISCUSSION TITLE
[SignalR] InvokeAsync<> hang in OnConnectedAsync on aspnet core (net7) Signalr

### DISCUSSION THREAD

**Discussion by maiemy** [dotnet/aspnetcore#48515] [Q&A]
Ciao to all,

I have a problem with InvokeAsync<> when I call it inside OnConnectedAsync in my hub implementation

        public override async Task OnConnectedAsync()
        {
            string connId = Context.ConnectionId;
            await Groups.AddToGroupAsync(connId, "SignalR Users");
            await _mia.Hub.Clients.All.SendAsync("NewComers", connId);
            string risultato = await _mia.Hub.Clients.Client(connId).InvokeAsync<string>("NewComerOnly", $"Da OnConnected Benvenuto {connId}", new CancellationToken());
            await base.OnConnectedAsync();
        }

Hub.Clients.All.SendAsync("NewComers", connId); works fine
Hub.Clients.Client(connId).InvokeAsync<string> send correctly the message to the client, but hang for reply and not receive string result from client (no exceptions raised from server)

client code:

        connection.on("NewComerOnly", (mgs) => {
            console.log(mgs);
            return "Grazie!";
        });

If I do the same in separated method

        public async Task<string> MioMetodo(string valore)
        {
            string connId = Context.ConnectionId;
            string risultato = string.Empty ;
            try
            {
                 risultato = await _mia.Hub.Clients.Client(connId).InvokeAsync<string>("NewComerOnly", $"Benvenuto {connId}", new CancellationToken());
            }catch(Exception ex)
            {
                var te = ex;
            }
            return risultato;
        }

all work fine.

Could you please help me to understand why?

ciao

**Comment by gfoidl** [comment]
Does [AspNetCore.SignalR SendAsync not firing inside OnConnectedAsync](https://stackoverflow.com/questions/54332558/aspnetcore-signalr-sendasync-not-firing-inside-onconnectedasync) apply here?

**Reply by maiemy** [reply]
> Does [AspNetCore.SignalR SendAsync not firing inside OnConnectedAsync](https://stackoverflow.com/questions/54332558/aspnetcore-signalr-sendasync-not-firing-inside-onconnectedasync) apply here?

thanks for your reply.
I could be wrong but this is not the case because I'm not using a Strongly Typed Hub
I'll copy the Hub for you to see what it looks like.



    public class TestHub : Hub
    {
        IHubContext _contextHub;
        MiaClasseTest _mia;

        public TestHub(IMiaClasseTest mia)
        {            
            _mia = (MiaClasseTest)mia;
        } // chiudo public TestHub(MiaClasseTest mia)

        public override async Task OnConnectedAsync()
        {
            string connId = Context.ConnectionId;
            await Groups.AddToGroupAsync(connId, "SignalR Users");
            await _mia.Hub.Clients.All.SendAsync("NewComers", connId);
            string risultato = await _mia.Hub.Clients.Client(connId).InvokeAsync<string>("NewComerOnly", $"Da OnConnected Benvenuto {connId}", new CancellationToken());
            await base.OnConnectedAsync();
        }
    }

Hub.Clients.All.SendAsync works fine 
alse Hub.Clients.Client(connId) send correctly but hang and not receive result from client.

**Reply by gfoidl** [reply]
@BrennanConroy can you please have a look here?

**Reply by BrennanConroy** [reply]
`InvokeAsync` does not work inside `OnConnectedAsync`, we made the common scenario of using `Context.Clients.Client(...).InvokeAsync` throw, but you're doing an unusual call pattern to get around that check.

`_mia.Hub.Clients.Client(connId)` is the unusual call pattern. Even without involving `InvokeAsync` it looks suspicious. What is `_mia.Hub`? If it's a `Hub` instance then you're doing something wrong, `Hub` instances should not be stored or created by the user.

**Reply by maiemy** [reply]
> `InvokeAsync` does not work inside `OnConnectedAsync`, we made the common scenario of using `Context.Clients.Client(...).InvokeAsync` throw, but you're doing an unusual call pattern to get around that check.
> 
> `_mia.Hub.Clients.Client(connId)` is the unusual call pattern. Even without involving `InvokeAsync` it looks suspicious. What is `_mia.Hub`? If it's a `Hub` instance then you're doing something wrong, `Hub` instances should not be stored or created by the user.

_mia is a simple test class that contains the instance of IHubContext 
at startup time I do an AddSingleton on services initializing MiaClasseTest class passing the instance of HubContext taken from Services.
So _mia.Hub is the instance of HubContext coming from Services.

_mia.Hub isn't a instance of Hub but IHubContext.


at this time I do not have sourcecode available. I will post it to you as soon as possible.

**Reply by BrennanConroy** [reply]
Ah, well my previous statement still stands. You can't call InvokeAsync from OnConnectedAsync.

**Reply by maiemy** [reply]
> Ah, well my previous statement still stands. You can't call InvokeAsync from OnConnectedAsync.

thank you.
can i ask you why it doesn't work? it's just my technical curiosity. :-)

**Reply by BrennanConroy** [reply]
We don't start the connections receive loop until `OnConnectedAsync` completes, so if you called `InvokeAsync`, it's waiting for a message from the client, but the receive loop isn't running yet, so it can't receive the message it's waiting for. If you used `Context.Clients.Client(...).InvokeAsync` we can detect that and throw an error. But using a random `IHubContext` the comes from outside of the Hub can't be detected sadly.

This is tracked in https://github.com/dotnet/aspnetcore/issues/41997#issuecomment-1222574613, the only idea I have so far is an analyzer that will try to detect this at build time.

**Reply by maiemy** [reply]
> We don't start the connections receive loop until `OnConnectedAsync` completes, so if you called `InvokeAsync`, it's waiting for a message from the client, but the receive loop isn't running yet, so it can't receive the message it's waiting for. If you used `Context.Clients.Client(...).InvokeAsync` we can detect that and throw an error. But using a random `IHubContext` the comes from outside of the Hub can't be detected sadly.
> 
> This is tracked in [#41997 (comment)](https://github.com/dotnet/aspnetcore/issues/41997#issuecomment-1222574613), the only idea I have so far is an analyzer that will try to detect this at build time.

thank you very much for your explanation.
I used this solution to notify the client of activation by passing a specific parameter when connecting with the server.

at server startup I use this code for take a reference of HubContext in my MiaClasseTest

            builder.Services.AddSingleton<IMiaClasseTest>(
                    iSerProvider => {
                         IHubContext<TestHub> hubContext = iSerProvider.GetRequiredService<IHubContext<TestHub>>();
                        return new MiaClasseTest("PIPPO!!", hubContext);
                    });

I made this choice in order to always have hubcontext available to communicate to specific clients, because in production I will have a timer that will send data to clients.

thank you very much again for your explanations
